DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/fft2d/src/fft2d_demo.cpp

```diff
@@ -0,0 +1,420 @@
+#endif
+
+      input_data = sycl::malloc_host<ac_complex<float>>(kN * kN, q);
+      output_data = sycl::malloc_host<ac_complex<float>>(kN * kN, q, prop_list);
+      temp_data = sycl::malloc_host<ac_complex<float>>(kN * kN, q, prop_list);
+    } else {
+      std::cerr << "USM device allocations or USM host allocations must be "
+                   "supported to run this sample."
+                << std::endl;
+      std::terminate();
+    }
+
+    if (input_data == nullptr || output_data == nullptr ||
+        temp_data == nullptr) {
+      std::cerr << "Failed to allocate USM memory." << std::endl;
+      std::terminate();
+    }
+
+    // Copy the input data from host DDR to USM memory
+    q.memcpy(input_data, host_input_data, sizeof(ac_complex<float>) * kN * kN)
+        .wait();
+
+    std::cout << "Launching a " << kN * kN << " points " << kParallelism
+              << "-parallel " << (inverse ? "inverse " : "")
+              << "FFT transform (" << (mangle ? "alternative" : "ordered")
+              << " data layout)" << std::endl;
+
+    /*
+     * A 2D FFT transform requires applying a 1D FFT transform to each matrix
+     * row followed by a 1D FFT transform to each column of the intermediate
+     * result.
+     * A single FFT engine can process rows and columns back-to-back. However,
+     * as matrix data is stored in global memory, the efficiency of memory
+     * accesses will impact the overall performance. Accessing consecutive
+     * memory locations leads to efficient access patterns. However, this is
+     * obviously not possible when accessing both rows and columns.
+     *
+     * The implementation is divided between three concurrent SYCL kernels, as
+     * depicted below:
+     *
+     *  --------------------      --------------      --------------------------
+     *  | read matrix rows | ---> | FFT engine | ---> | bit-reverse, transpose |
+     *  |                  |      |            |      |    and write matrix    |
+     *  --------------------      --------------      --------------------------
+     *
+     * This sequence of kernels does back-to-back row processing followed by a
+     * data transposition and writes the results back to memory. The host code
+     * runs these kernels twice to produce the overall 2D FFT transform
+     *
+     *
+     * These kernels transfer data through pipes.
+     * This avoids the need to read and write intermediate data using global
+     * memory.
+     *
+     * In many cases the FFT engine is a building block in a large application.
+     * In this case, the memory layout of the matrix can be altered to achieve
+     * higher memory transfer efficiency. This implementation demonstrates how
+     * an alternative memory layout can improve performance. The host switches
+     * between the two memory layouts using a kernel argument. See the
+     * 'MangleBits' function for additional details.
+     */
+
+    double start_time{};
```

Comment:
Why the empty brackets?    If you want to specifically initialize the variable to a value you should specify the value.

Reply:
True, removing them